src/Data/Expr.purs

```diff
@@ -2,89 +2,389 @@ module Lynx.Data.Expr where
 
 import Prelude
 
+import Control.Alt ((<|>))
 import Data.Argonaut (class DecodeJson, class EncodeJson, Json, decodeJson, encodeJson, jsonEmptyObject, stringify, (.:), (:=), (~>))
 import Data.BigInt as Data.BigInt
 import Data.DateTime (DateTime)
 import Data.Either (Either(..))
-import Data.Foldable (fold, intercalate)
+import Data.Eq (class Eq1)
+import Data.Foldable (class Foldable, fold, foldlDefault, foldrDefault)
 import Data.Formatter.DateTime (format, unformat)
 import Data.Formatter.Parser.Interval (extendedDateTimeFormatInUTC)
+import Data.Functor.Coproduct.Inject (inj, prj)
+import Data.Functor.Coproduct.Nested (type (<\/>), (<\/>))
+import Data.Functor.Mu (Mu)
 import Data.Generic.Rep (class Generic)
 import Data.Generic.Rep.Show (genericShow)
-import Data.Maybe (Maybe(..), maybe')
-import Data.Newtype (wrap)
+import Data.Maybe (Maybe(..), maybe, maybe')
+import Data.Newtype (class Newtype, un, wrap)
 import Data.NonEmpty (NonEmpty(..))
 import Data.String.Common (replace)
 import Data.String.Pattern (Pattern(..), Replacement(..))
+import Data.Traversable (class Traversable, traverse)
 import Foreign.Object (Object)
 import Foreign.Object as Foreign.Object
+import Matryoshka (class Corecursive, class Recursive, anaM, cata, embed, project)
 import Ocelot.Data.Currency (Cents, formatCentsToStrDollars)
 import Test.QuickCheck (class Arbitrary, arbitrary)
 import Test.QuickCheck.Gen (Gen, Size, arrayOf, oneOf, sized)
 
 type Key = String
 
-data ExprType
-  = Array (Array ExprType)
-  | Boolean Boolean
-  | Cents Cents
-  | DateTime DateTime
-  | Int Int
-  | Pair { name :: ExprType, value :: ExprType }
-  | String String
+newtype ArrayF a
+  = Array (Array a)
+
+derive instance eqArrayF :: (Eq a) => Eq (ArrayF a)
+
+derive instance eq1ArrayF :: Eq1 ArrayF
+
+derive instance functorArrayF :: Functor ArrayF
+
+derive newtype instance showArrayF :: (Show a) => Show (ArrayF a)
+
+derive newtype instance foldableArrayF :: Foldable ArrayF
+
+derive newtype instance traversableArrayF :: Traversable ArrayF
+
+decodeArrayF :: ExprJSON -> Either String (ArrayF ExprJSON)
+decodeArrayF { out, param } = case out of
+  "Array" -> map Array (decodeJson param)
+  _ -> Left "Expected Array"
+
+encodeArrayF :: ArrayF ExprJSON -> ExprJSON
+encodeArrayF = case _ of
+  Array x -> base "Array" x
+
+isEmptyArrayF :: forall a. ArrayF a -> Boolean
+isEmptyArrayF = case _ of
+  Array [] -> true
+  Array _ -> false
+
+printArrayF :: ArrayF String -> String
+printArrayF = case _ of
+  Array x -> show x
+
+newtype BooleanF a
+  = Boolean Boolean
+
+derive instance eqBooleanF :: Eq (BooleanF a)
+
+derive instance eq1BooleanF :: Eq1 BooleanF
+
+derive instance functorBooleanF :: Functor BooleanF
+
+derive newtype instance showBooleanF :: Show (BooleanF a)
+
+instance foldableBooleanF :: Foldable BooleanF where
+  foldl f z x = foldlDefault f z x
+  foldr f z x = foldrDefault f z x
+  foldMap f = case _ of
+    Boolean _ -> mempty
```

Comment:
Can't all of these just be
```purescript
foldMap f _ = mempty
```
Or maybe there's something I'm missing?

Reply:
You're not missing anything. I'm a judicious `case`er. If we're going for concision, we could do `foldMap = mempty`.

Reply:
I am sort of feeling the pattern matching to be unnecessary here. We don't actually care about the value or the constructor, in which case I think it makes sense to just `foldMap _ _ = mempty`